### PR TITLE
Apply new frontend design to Container and deriving classes

### PR DESCRIPTION
**Pull request by franzpoeschel**
Follow-up to #886.

This splits `Container` and deriving classes into a data and an interface class.

TODO:
- [x] Do this to all those classes. Already done: `Container`, `BaseRecord`, `RecordComponent`, `BaseRecordComponent`, `PatchRecordComponent`, `Iteration`
- [x] Don't split those classes as they don't have extra data members: `PatchRecord`, `ParticlePatches`, `ParticleSpecies` (its only member `ParticlePatches` already has handle semantics), `Mesh`
- [x] Fix Python bindings
- [x] Update documentation

This PR prepares for a necessary UX frontend redesign that will be a follow-up:

* The `SCALAR` trick for instantiating datasets that have no x,y,z components is confusing for users, I've had to explain this several times now and it doesn't really make sense. Make it so that instead of `auto E_x = series.iterations[0].meshes["temperature"][SCALAR]` one can equivalently write `auto E_x = series.iterations[0].meshes["temperature"]`. 
* Allow users to create custom subgroups

To do this, probably do something like:
* Separate the functionality to create subgroups into a class (maybe reuse the class `Record` for that as it has a similar purpose
* Separate the functionality to create datasets into a class (now part of `RecordComponent`)

### CODE REVIEW

**Comment by franzpoeschel** [comment]
I have now also simplified the whole design again a bit.

**First approach:** Split `Series` into `SeriesInterface` and `SeriesData`. Have two classes that put these two together:
* `internal::SeriesInternal`: This has a field `SeriesData data;` and simply adds `SeriesInterface` to it. Basically, nothing else but a `SeriesData` with an interface. Non-copiable, non-movable.
* `Series`: The public class. It has a field `std::shared_ptr<SeriesData> data;` and also adds `SeriesInterface` to it.

While this works alright, it's quite a lot of classes.
**Second approach:** I've tried to reduce it to two classes now:
* `internal::SeriesData`: unchanged compared to previous approach
* `Series`: Now contains all the logic of the `Series`, and a shared pointer to `internal::SeriesData`.

This is now quite close to the old implementation again, with one key difference: All stateful data held by the `Series` class is found at one object that stays at the same place in memory. So, our classes still have one unique internal address which was the goal of the whole redesign.

The class `internal::SeriesInternal` is now completely gone. A non-owning instance of `Series` can be created from an instance of `internal::SeriesData` by defining a `shared_ptr` without destructor:
```cpp
Series nonOwningInternalSeries{ std::shared_ptr< internal::SeriesData >{ &seriesData, []( auto const * ){} } };
```

One downside of this is that for class hierarchies, you now have one redundant `shared_ptr` per layer in the hierarchy. They are redundant, since they point to the same data, but `static_cast`ed to each layer in the hierarchy. There is still only one internal address per object.

**Comment by franzpoeschel** [comment]
I've added documentation for this now. Otherwise, this PR is finished from my side and ready for review @ax3l 
It's likely that #1025 will be slightly broken (`SeriesInterface` and `AttributableInterface` are now gone again)

**Comment by franzpoeschel** [comment]
@eschnett Hi Erik,
this PR will likely lead to conflicts with your Julia bindings: It simplifies the design of `Series` and `Attributable`, by e.g. removing the `SeriesInterface` class and retaining only two classes: `SeriesData` for all data members and `Series` for the implementation. Similar for `Attributable`.
The same separation between data and implementation classes is applied to several other classes in the openPMD hierarchy, see comments above.

How should we go forward on this? Would it be ok for you to check how much this breaks the Julia bindings? In which order would you prefer your and this PR to be applied?

There's no hurry as I'm in vacation next week.

(Also, some time after your PR is merged, we generally should synchronize on how to maintain the bindings)

**Comment by eschnett** [comment]
@franzpoeschel I think the ball is currently in my court to add CI tests for the Julia bindings. Before that it'd be difficult for you to maintain the Julia bindings, or to update them when necessary.

Given this, I'd say: go ahead and merge the changes. The branch with the Julia bindings will continue to work (since it doesn't have the new changes), and I'll merge the changes and update the Yggdrasil build in one go. This way there won't be any breakage seen from Julia.

I'd love to have a preview on your changes, but realistically not in the next week, so don't let me slow you down.

**Comment by franzpoeschel** [comment]
Ok, then we'll go forward with this first :)

If you need any help merging, just ask

File: include/openPMD/backend/BaseRecordComponent.hpp [hunk 1]
```diff
@@ -79,10 +119,34 @@ class BaseRecordComponent : public LegacyAttributable
...
+
+    inline internal::BaseRecordComponentData const & get() const
+    {
+        return *m_baseRecordComponentData;
+    }
+
+    inline internal::BaseRecordComponentData & get()
+    {
+        return const_cast< internal::BaseRecordComponentData & >(
+            static_cast< BaseRecordComponent const * >( this )->get() );
```
Comment (ax3l) [inline include/openPMD/backend/BaseRecordComponent.hpp:115]: For these getters, wouldn't `return *m_baseRecordComponentData;` work here as well and be more legible? :)

I think duplicating the single line does no harm as they are close by and is more readable then the current construct.

File: include/openPMD/RecordComponent.hpp [hunk 2]
```diff
@@ -78,6 +78,40 @@ struct IsContiguousContainer< std::array< T_Value, N > >
...
+
+        RecordComponentData();
+
+        RecordComponentData( RecordComponentData const & ) = delete;
+        RecordComponentData( RecordComponentData && ) = delete;
+
+        RecordComponentData & operator=( RecordComponentData const & ) = delete;
+        RecordComponentData & operator=( RecordComponentData && ) = delete;
+
+        std::queue< IOTask > m_chunks;
```
Comment (ax3l) [inline include/openPMD/RecordComponent.hpp:96]: For all the `...Data` classes, it would be really good if we try to catch up and add Doxygen strings to the `m_` member variables.

Can be a follow-up and we can split the work if we are unsure about some. I remember we found them hard to follow at times and it's good to document the purpose of individual `m_` member variables.

**Review by ax3l — COMMENTED** [review]
This looks really good to me. Thanks a lot for this!

Just one comment about the implementation of the `get()`ers and one follow-up idea about catching up documenting the `m_` members in `...Data` classes.

**Comment by ax3l** [comment]
Starting CI again because we haven't pushed to the branch in a bit (to see if merge is still green).

**Review by ax3l — APPROVED** [review]
If CI passes then I'll merge this already, since my comments can all be addressed in follow-ups.

File: include/openPMD/backend/BaseRecordComponent.hpp (hunk 1 above)
Comment (franzpoeschel) [reply include/openPMD/backend/BaseRecordComponent.hpp:115]: Yeah .. maybe you're right about this :D 
bit of tunnel vision on my side there

File: include/openPMD/RecordComponent.hpp (hunk 2 above)
Comment (franzpoeschel) [reply include/openPMD/RecordComponent.hpp:96]: Note: Inside #1154 I began aliasing these data classes (needed that for something else in there) which made the whole thing a lot more readable in my opinion, e.g.:
```cpp
namespace internal {
    class RecordComponentData{ /* … */ };
}

class RecordComponent {
    using DataClass = internal::RecordComponentData;
    std::shared_ptr< DataClass > m_recordComponentData;
    /* … */
};
```

File: include/openPMD/backend/BaseRecordComponent.hpp (hunk 1 above)
Comment (ax3l) [reply include/openPMD/backend/BaseRecordComponent.hpp:115]: No worries, feel free to fix in a separate follow-up.

File: include/openPMD/RecordComponent.hpp (hunk 2 above)
Comment (ax3l) [reply include/openPMD/RecordComponent.hpp:96]: That is very nice, yes :sparkling_heart: 

But my comment was actually about adding doxygen to the members in the data class, so we know what they are for:
```C++
struct RecordComponentData
{
    // ...

    //! container for holding data chunks
    std::queue< IOTask > m_chunks;

    //! this sets things on fire
    bool m_fire;
};
```

I recall that all the original `m_dirty` and `m_flushed` members among others were often hard to understand for us without searching through their usage.

Comment (franzpoeschel) [reply include/openPMD/RecordComponent.hpp:96]: Ah, got you
Yep, these classes are probably a good chance to have documentation for our internal state at one central place
